src/app/claim/page.tsx

```diff
@@ -7,13 +7,18 @@ import {
   getTree,
 } from "@/lib/merkle/merkle";
 import { getAddressForSession, getServerSession } from "@/lib/session";
+import { redirect } from "next/navigation";
 import { formatUnits } from "viem";
 
 export default async function Claim() {
   const session = await getServerSession();
+  if (!session?.isKycVerified && !session?.hasClaimed) {
+    redirect("/");
+  }
+
   const fullAddress = getAddressForSession(session);
-  const allocation = getAllocationForAddress(fullAddress);
-  const merkleProof = getProofForAddress(fullAddress, getTree());
+  const allocation = await getAllocationForAddress(fullAddress);
```

Comment:
Consider `useMemo` or something similar as this in theory would be triggered & would be blocking more often than needed